json-path/src/main/java/com/jayway/jsonpath/internal/function/MapJsonPathFunctionFactory.java

```diff
@@ -0,0 +1,30 @@
+package com.jayway.jsonpath.internal.function;
+
+import com.jayway.jsonpath.InvalidPathException;
+
+import java.util.Map;
+
+public class MapJsonPathFunctionFactory implements JsonPathFunctionFactory {
+
+    private final Map<String, Class<? extends PathFunction>> pathFunctionMap;
+
+    public MapJsonPathFunctionFactory(
+            final Map<String, Class<? extends PathFunction>> pathFunctionMap) {
+
+        this.pathFunctionMap = pathFunctionMap;
+    }
+
+    @Override
+    public PathFunction newFunction(final String name) throws InvalidPathException {
```

Comment:
I question whether PathFunction as an interface is an appropriate expression for a user facing function feature - there's some nuance to know in what context the function is being invoked that is internal state of the JsonPath library.  Breaking apart the types of PathFunctions (aggregations vs straight calculations) might be a good way to tease apart the interface into something more meaningful for the end-user.  

Also, does the function need to take parameters -- can a sub-select be a parameter to the function - expressing that in a more general way such that the consumer doesn't need to grok the internals of JsonPath would be ideal.

Certainly keep this method as all interfaces/implementations are a PathFunction, but I think you might want to introduce additional factory methods for aggregation vs straight calculation and then abstract away the nuance of accepting parameters.